### PR TITLE
s3 reader first cut

### PR DESCRIPTION
**Pull request by macgyver603**
First cut at an S3 reader using the modern TS processor APIs

This PR will also need:  
- [x] S3 writer  
- [x] Test coverage

### CODE REVIEW

File: asset/s3_exporter/processor.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            throw new Error(`Unsupported output format "${this.opConfig.format}"`);
+        }
+
+        const params = {
+            Bucket: this.opConfig.bucket,
+            Key: objName,
+            Body: outStr
+        };
+
+        return this.client.putObject_Async(params);
```
Comment (peterdemartini) [inline asset/s3_exporter/processor.js]: We should be careful about what we are returning from this processor, it should be an array of objects or data entities.

File: asset/s3_reader/fetcher.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        this._shutdown = true;
+        return super.shutdown();
+    }
+
+    async fetch(slice) {
+        // Coerce the field delimiter if the format is `tsv`
+        if (this.opConfig.format === 'tsv') {
+            this.opConfig.field_delimiter = '\t';
+        }
+        this.reader = (offset, length) => {
```
Comment (peterdemartini) [inline asset/s3_reader/fetcher.js]: We shouldn't be assigning this function to `this`. 
```suggestion
        const reader = (offset, length) => {
```

File: test/s3_reader/slicer-spec.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        // Make sure the slicer has time to slice
+        setTimeout(() => {
+            expect(slicerJSON._queue._size).toEqual(6);
+            expect(slicerJSON._doneSlicing).toEqual(true);
+            expect(slicerJSON._lastKey).toEqual('obj2');
+        }, 1000);
+    });
+
+
+    it('slices non-JSON objects correctly', async () => {
```
Comment (peterdemartini) [inline test/s3_reader/slicer-spec.js]: The `it` block will resolve before the code will finish.

File: test/s3_reader/slicer-spec.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        },
+        {
+            name: 's3_exporter'
+        });
+
+    // Set the clients to our test client
+    slicerJSON.client = mockClient;
+    slicerOther.client = mockClient;
+
+    it('slices JSON objects into the expected number of slices.', async () => {
```
Comment (peterdemartini) [inline test/s3_reader/slicer-spec.js]: The `it` block will resolve before the code will finish.

File: asset/s3_exporter/processor.js (hunk 1 above)
Comment (macgyver603) [reply asset/s3_exporter/processor.js]: Changed this to return a data entity with the results of `putObject_Async()`

File: asset/s3_reader/fetcher.js (hunk 2 above)
Comment (macgyver603) [reply asset/s3_reader/fetcher.js]: fixed

File: test/s3_reader/slicer-spec.js (hunk 3 above)
Comment (macgyver603) [reply test/s3_reader/slicer-spec.js]: fixed with the `test-harness`

File: test/s3_reader/slicer-spec.js (hunk 4 above)
Comment (macgyver603) [reply test/s3_reader/slicer-spec.js]: fixed with the `test-harness`

**Pull request merged**